vault/seal_autoseal_test.go

```diff
@@ -0,0 +1,151 @@
+package vault
+
+import (
+	"bytes"
+	"context"
+	"reflect"
+	"testing"
+
+	proto "github.com/golang/protobuf/proto"
+	"github.com/hashicorp/vault/sdk/physical"
+	"github.com/hashicorp/vault/vault/seal"
+)
+
+type phy struct {
+	t       *testing.T
+	entries map[string][]*physical.Entry
+}
+
+func newTestBackend(t *testing.T) *phy {
+	return &phy{
+		t:       t,
+		entries: make(map[string][]*physical.Entry),
+	}
+}
+
+func (p *phy) Put(_ context.Context, entry *physical.Entry) error {
+	p.entries[entry.Key] = append(p.entries[entry.Key], entry)
+	return nil
+}
+
+func (p *phy) Get(_ context.Context, key string) (*physical.Entry, error) {
+	entries := p.entries[key]
+	if entries == nil {
+		return nil, nil
+	}
+	return entries[len(entries)-1], nil
+}
+
+func (p *phy) Delete(_ context.Context, key string) error {
+	p.t.Errorf("Delete called on phy: key: %v", key)
+	return nil
+}
+
+func (p *phy) List(_ context.Context, prefix string) ([]string, error) {
+	p.t.Errorf("List called on phy: prefix: %v", prefix)
+	return []string{}, nil
+}
+
+func (p *phy) Len() int {
+	return len(p.entries)
+}
+
+func TestAutoSeal_UpgradeKeys(t *testing.T) {
+	core, _, _ := TestCoreUnsealed(t)
+	testSeal := seal.NewTestSeal(nil)
+
+	var encKeys []string
+	changeKey := func(key string) {
+		encKeys = append(encKeys, key)
+		testSeal.SetKeyID(key)
+	}
+
+	// Set initial encryption key.
+	changeKey("kaz")
+
+	autoSeal := NewAutoSeal(testSeal)
+	autoSeal.SetCore(core)
+	pBackend := newTestBackend(t)
+	core.physical = pBackend
+
+	ctx := context.Background()
+
+	inkeys := [][]byte{[]byte("grist"), []byte("house")}
```

Comment:
I would use a single stored key, as that's really all that's supported in practice (and once my 1.3 changes get merged, it will be enforced.)

Reply:
The autoSeal code currently allows for multiple stored keys which is why I used multiple keys in the test. I will update the test if your PR gets merged before this one. (Assuming your PR breaks these tests). Can you add a reference to your PR in a comment so we can link them?

Reply:
In theory it allows it, but all the real autoseal implementations never store more than one.

Here's the PR: https://github.com/hashicorp/vault-enterprise/pull/1034